src/java/org/apache/cassandra/schema/DistributedSchema.java

```diff
@@ -114,16 +114,29 @@ public KeyspaceMetadata getKeyspaceMetadata(String keyspace)
     public static DistributedSchema fromSystemTables(Keyspaces keyspaces, Set<String> knownDatacenters)
     {
         if (!keyspaces.containsKeyspace(SchemaConstants.METADATA_KEYSPACE_NAME))
-            keyspaces = keyspaces.withAddedOrReplaced(Keyspaces.of(DistributedMetadataLogKeyspace.initialMetadata(knownDatacenters),
-                                                                   TraceKeyspace.metadata(),
-                                                                   SystemDistributedKeyspace.metadata(),
-                                                                   AuthKeyspace.metadata()));
+        {
+            Keyspaces kss = Keyspaces.of(DistributedMetadataLogKeyspace.initialMetadata(knownDatacenters),
+                                         TraceKeyspace.metadata(),
+                                         SystemDistributedKeyspace.metadata(),
+                                         AuthKeyspace.metadata());
+            for (KeyspaceMetadata ksm : keyspaces) // on disk keyspaces
+                kss = kss.withAddedOrUpdated(kss.get(ksm.name)
+                                                .map(k -> merged(k, ksm))
+                                                .orElse(ksm));
+            keyspaces = kss;
+        }
         return new DistributedSchema(keyspaces, Epoch.UPGRADE_GOSSIP);
     }
 
-    public void initializeKeyspaceInstances(DistributedSchema prev)
+    private static KeyspaceMetadata merged(KeyspaceMetadata ksm1, KeyspaceMetadata ksm2)
```

Comment:
nit: would it make sense to name these `newKs/oldKs` or similar? Not a big deal really as there's only the one usage of the method for now.